pkg/kv/kvtest/store.go

```diff
@@ -369,3 +374,310 @@ func testScanPrefix(t *testing.T, ms MakeStore) {
+
+	// Using the following keys (for some ints:L,M,N, and unique TESTID), that will provide access
+
+	// TESTID-pref-key-0000
+	// TESTID-pref-key-0001
+	// ...
+	// TESTID-pref-key-000<N>
+	// TESTID-prefin-key-0000
+	// TESTID-prefin-key-0001
+	// ...
+	// TESTID-prefin-key-000<M>
+	// TESTID-prefinin-key-0000
+	// TESTID-prefinin-key-0001
+	// ...
+	// TESTID-prefinin-key-000<L>
+
+	extPref := uniqueKey("pref")
+	inPref := uniqueKey("prefin")
+	ininPref := uniqueKey("prefinin")
+
+	allPrefs := [][]byte{extPref, inPref, ininPref}
+
+	// delete a subsection of the read prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, inPref, ininPref)
+
+	// scan a subsection of the delete prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, ininPref, inPref)
+
+	// scan and delete the same prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, ininPref, ininPref)
+}
+
+func testDeleteWhileIterSamePrefixSingleRun(t *testing.T, ms MakeStore, prefsToCreate [][]byte, readPref, delPref []byte) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	// Emptying the KV store from previous entries. This is essential in these tests as
+	// the verification relies on exact keys to exist
+	cleanIter, err := store.Scan(ctx, delPref)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanIter.Close()
+
+	for cleanIter.Next() {
+		e := cleanIter.Entry()
+		if cleanIter != nil {
+			err = store.Delete(ctx, e.Key)
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
+
+	for _, pref := range prefsToCreate {
+		setupSampleData(t, ctx, store, string(pref), 100)
+	}
+
+	// Will be filled by the scan&read routine, to later verify that all values are passed by scan
+	readDone := make(map[string]bool)
+	deleteDone := make(map[string]bool)
+
+	errGrp, errCtx := errgroup.WithContext(context.Background())
+
+	// Scan and read
+	errGrp.Go(func() error {
+		ei, err := store.Scan(errCtx, readPref)
+		if err != nil {
+			return fmt.Errorf("unexpected error from store.Scan (read): %w", err)
+		}
+		defer ei.Close()
+
+		for ei.Next() {
+			e := ei.Entry()
+			if e != nil {
+				readDone[string(e.Key)] = true
+			} else if ei.Err() != nil {
+				return fmt.Errorf("unexpected error during read iteration: %w", err)
+			}
+		}
+		return nil
+	})
+
+	// Scan and delete
+	errGrp.Go(func() error {
+		ei, err := store.Scan(errCtx, delPref)
+		if err != nil {
+			return fmt.Errorf("unexpected error from store.Scan (delete): %w", err)
+		}
+		defer ei.Close()
+
+		for ei.Next() {
+			e := ei.Entry()
+			if e == nil {
+				return fmt.Errorf("unexpected nil entry during deletion interation: %w", ei.Err())
+			}
+			err = store.Delete(errCtx, e.Key)
+			if err != nil {
+				return fmt.Errorf("unexpected delete failure :%w", err)
+			}
+			deleteDone[string(e.Key)] = true
+		}
+		return nil
+	})
+
+	err = errGrp.Wait()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// verify all readDone entries fits readPref
+	for k := range readDone {
```

Comment:
We should add another check (verify all readPref has readDone) that scans the prefix again - and verifies that each entry has a true value in readDone

Reply:
It is not required that all entries under `readPref` will be read. Since we are reading and deleting overlapping prefixes (purpose of this test), an entry that satisfy both `delPref` and `readPref`can be deleted before it is read. We do need to guarantee that if an entry was read, it satisfies `readPref`, but not the other way around

Reply:
What I meant was that we need to make sure that all the entries in the range were either deleted or read. What you checked here is that everything you read, wasn't deleted which is another thing.

Reply:
Done